fdbserver/MoveKeys.actor.cpp

```diff
@@ -373,6 +374,192 @@ ACTOR Future<Optional<UID>> checkReadWrite(Future<ErrorOr<GetShardStateReply>> f
+	tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+	loop {
+		RangeResult readResult = wait(krmGetRanges(tr,
+		                                           serverKeysPrefixFor(ssid),
+		                                           KeyRangeRef(toReadRangeBegin, range.end),
+		                                           CLIENT_KNOBS->KRM_GET_RANGE_LIMIT,
+		                                           CLIENT_KNOBS->KRM_GET_RANGE_LIMIT_BYTES));
+		// We do not want to reset tr
+		for (int i = 0; i < readResult.size() - 1; i++) {
+			UID shardId;
+			bool assigned, emptyRange;
+			EnablePhysicalShardMove enablePSM = EnablePhysicalShardMove::False;
+			decodeServerKeysValue(readResult[0].value, assigned, emptyRange, enablePSM, shardId);
+			if (!assigned) {
+				TraceEvent(SevError, "ValidateRangeAssignmentError")
+				    .detail("Context", context)
+				    .detail("AuditRange", range)
+				    .detail("ErrorMessage", "KeyServers has range but ServerKeys does not have")
+				    .detail("CurrentEmptyRange", emptyRange)
+				    .detail("CurrentAssignment", assigned)
+				    .detail("EnablePSM", enablePSM)
+				    .detail("ServerID", ssid)
+				    .detail("ShardID", shardId);
+				allCorrect = false;
+			}
+		}
+		if (readResult.back().key < range.end) {
+			toReadRangeBegin = readResult.back().key;
+			TraceEvent(SevWarnAlways, "ValidateRangeAssignmentMultipleReads")
+			    .detail("Range", range)
+			    .detail("StorageServer", ssid);
+			continue;
+		} else {
+			break;
+		}
+	}
+	return allCorrect;
+}
+
+ACTOR Future<Void> auditLocationMetadataPreCheck(Transaction* tr,
+                                                 KeyRange range,
+                                                 std::vector<UID> servers,
+                                                 std::string context) {
+	if (range.empty()) {
+		TraceEvent(SevWarn, "AuditLocationMetadataPreCheckEmptyInputRange").detail("AuditRange", range);
+		return Void();
+	}
+	state std::vector<Future<Void>> actors;
+	state std::unordered_map<UID, bool> results;
+	state bool anyError = false;
+	state double startTime = now();
+	state int retryCount = 0;
+	try {
+		loop {
+			try {
+				actors.clear();
+				results.clear();
+				for (auto& ssid : servers) {
+					actors.push_back(store(results[ssid], validateRangeAssignment(tr, range, ssid, context)));
+				}
+				wait(waitForAllReadyThenThrow(actors));
+				for (const auto& [ssid, res] : results) {
+					if (!res)
+						anyError = true;
+				}
+				if (anyError) {
+					throw location_metadata_corruption();
+				}
+				break;
+			} catch (Error& e) {
+				if (e.code() == error_code_actor_cancelled) {
+					throw e;
+				}
+				if (now() - startTime >= 0.5) {
+					TraceEvent(SevWarn, "AuditLocationMetadataPreCheckTimedOut")
+					    .detail("AuditRange", range)
+					    .detail("Servers", describe(servers))
+					    .detail("RetryCount", retryCount);
+					throw timed_out();
+				}
+				retryCount++;
+				wait(delay(0.1));
```

Comment:
ditto: is this timeout necessary? Will it be converted to `audit_storage_failed` in the following catch?